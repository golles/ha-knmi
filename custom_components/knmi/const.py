"""Constants for knmi."""
from datetime import timedelta
from typing import Final

# API
API_ENDPOINT: Final = "http://weerlive.nl/api/json-data-10min.php?key={}&locatie={},{}"
API_TIMEOUT: Final = 10
API_TIMEZONE: Final = "Europe/Amsterdam"

# Base component constants.
NAME: Final = "KNMI"
DOMAIN: Final = "knmi"
VERSION: Final = "1.3.7"

# Defaults
DEFAULT_NAME: Final = NAME
SCAN_INTERVAL = timedelta(seconds=300)

# Platforms.
BINARY_SENSOR: Final = "binary_sensor"
SENSOR: Final = "sensor"
WEATHER: Final = "weather"
PLATFORMS: Final = [BINARY_SENSOR, SENSOR, WEATHER]
