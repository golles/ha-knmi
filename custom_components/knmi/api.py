"""KnmiApiClient"""
import asyncio
import logging
import socket

import aiohttp
import async_timeout

from .const import API_ENDPOINT, API_TIMEOUT

_LOGGER: logging.Logger = logging.getLogger(__package__)


class KnmiApiClient:
    """KNMI API wrapper"""

    def __init__(
        self,
        api_key: str,
        latitude: str,
        longitude: str,
        session: aiohttp.ClientSession,
    ) -> None:
        """Sample API Client."""
        self.api_key = api_key
        self.latitude = latitude
        self.longitude = longitude
        self._session = session

    async def async_get_data(self) -> dict:
        """Get data from the API."""
        url = API_ENDPOINT.format(self.api_key, self.latitude, self.longitude)
        return await self.api_wrapper("get", url)

    async def api_wrapper(self, method: str, url: str) -> dict:
        """Get information from the API."""
        try:
            async with async_timeout.timeout(API_TIMEOUT):
                if method == "get":
                    response = await self._session.get(url)
                    # The API has no proper error handling for a wrong API key.
                    # Instead a 200 with a message is returned, try to detect that here.
                    if "Vraag eerst een API-key op" in await response.text():
                        raise KnmiApiException("Invalid API key")

                    if "Dagelijkse limiet" in await response.text():
                        raise KnmiApiException("Exceeded Daily Limit")

                    data = await response.json()
                    return data.get("liveweer")[0]

        except asyncio.TimeoutError as exception:
            _LOGGER.error(
                "Timeout error fetching information from %s - %s",
                url,
                exception,
            )

        except (KeyError, TypeError) as exception:
            _LOGGER.error(
                "Error parsing information from %s - %s",
                url,
                exception,
            )
        except (aiohttp.ClientError, socket.gaierror) as exception:
            _LOGGER.error(
                "Error fetching information from %s - %s",
                url,
                exception,
            )
        except KnmiApiException as exception:
            _LOGGER.error("Error in API! - %s", exception)
            # Raise to pass on to the user.
            raise exception
        except Exception as exception:  # pylint: disable=broad-except
            _LOGGER.error("Something really wrong happened! - %s", exception)


class KnmiApiException(Exception):
    """KNMI API Exception class"""
