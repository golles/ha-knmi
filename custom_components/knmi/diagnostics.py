"""Diagnostics support for knmi."""
from homeassistant.components.diagnostics import async_redact_data
from homeassistant.config_entries import ConfigEntry
from homeassistant.const import CONF_API_KEY, CONF_LATITUDE, CONF_LONGITUDE
from homeassistant.core import HomeAssistant

from . import DOMAIN, KnmiDataUpdateCoordinator

TO_REDACT = {CONF_API_KEY, CONF_LATITUDE, CONF_LONGITUDE}


async def async_get_config_entry_diagnostics(
    hass: HomeAssistant, config_entry: ConfigEntry
) -> dict:
    """Return diagnostics for a config entry."""
    coordinator: KnmiDataUpdateCoordinator = hass.data[DOMAIN][config_entry.entry_id]

    return {
        "config_entry": async_redact_data(config_entry.as_dict(), TO_REDACT),
        "data": coordinator.data,
    }
