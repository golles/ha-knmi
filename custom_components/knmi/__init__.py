"""
Custom integration to integrate knmi with Home Assistant.

For more details about this integration, please refer to
https://github.com/golles/ha-knmi/
"""
import logging
from typing import Any, Callable

from homeassistant.config_entries import ConfigEntry
from homeassistant.const import CONF_API_KEY, CONF_LATITUDE, CONF_LONGITUDE
from homeassistant.core import Config, HomeAssistant
from homeassistant.exceptions import ConfigEntryNotReady
from homeassistant.helpers.aiohttp_client import async_get_clientsession
from homeassistant.helpers.device_registry import DeviceEntryType
from homeassistant.helpers.entity import DeviceInfo
from homeassistant.helpers.update_coordinator import DataUpdateCoordinator, UpdateFailed

from .api import KnmiApiClient
from .const import DOMAIN, NAME, PLATFORMS, SCAN_INTERVAL, VERSION

_LOGGER: logging.Logger = logging.getLogger(__package__)


async def async_setup(_hass: HomeAssistant, _config: Config) -> bool:
    """Set up this integration using YAML is not supported."""
    return True


async def async_setup_entry(hass: HomeAssistant, entry: ConfigEntry) -> bool:
    """Set up KNMI as config entry."""
    if hass.data.get(DOMAIN) is None:
        hass.data.setdefault(DOMAIN, {})

    api_key = entry.data.get(CONF_API_KEY)
    latitude = entry.data.get(CONF_LATITUDE)
    longitude = entry.data.get(CONF_LONGITUDE)

    session = async_get_clientsession(hass)
    client = KnmiApiClient(api_key, latitude, longitude, session)

    device_info = DeviceInfo(
        entry_type=DeviceEntryType.SERVICE,
        identifiers={(DOMAIN, entry.entry_id)},
        manufacturer=NAME,
        name=NAME,
        model=VERSION,
        configuration_url="http://weerlive.nl/api/toegang/account.php",
    )

    coordinator = KnmiDataUpdateCoordinator(
        hass=hass, client=client, device_info=device_info
    )
    await coordinator.async_config_entry_first_refresh()

    if not coordinator.last_update_success:
        raise ConfigEntryNotReady

    entry.async_on_unload(entry.add_update_listener(async_reload_entry))

    hass.data.setdefault(DOMAIN, {})[entry.entry_id] = coordinator

    for platform in PLATFORMS:
        if entry.options.get(platform, True):
            await hass.async_add_job(
                hass.config_entries.async_forward_entry_setup(entry, platform)
            )
    return True


async def async_unload_entry(hass: HomeAssistant, entry: ConfigEntry) -> bool:
    """Unload KNMI config entry."""
    unload_ok = await hass.config_entries.async_unload_platforms(entry, PLATFORMS)
    if unload_ok:
        del hass.data[DOMAIN][entry.entry_id]
    return unload_ok


async def async_reload_entry(hass: HomeAssistant, entry: ConfigEntry) -> None:
    """Reload config entry."""
    await async_unload_entry(hass, entry)
    await async_setup_entry(hass, entry)


class KnmiDataUpdateCoordinator(DataUpdateCoordinator):
    """Class to manage fetching data from the API."""

    def __init__(
        self, hass: HomeAssistant, client: KnmiApiClient, device_info: DeviceInfo
    ) -> None:
        """Initialize."""
        self.api = client
        self.device_info = device_info

        super().__init__(
            hass=hass, logger=_LOGGER, name=DOMAIN, update_interval=SCAN_INTERVAL
        )

    async def _async_update_data(self) -> dict[str, Any]:
        """Update data via library."""
        try:
            return await self.api.async_get_data()
        except Exception as exception:
            raise UpdateFailed() from exception

    def get_value(
        self, key: str, convert_to: Callable = str
    ) -> float | int | str | None:
        """Get a value from the retrieved data and convert to given type"""
        if self.data and key in self.data:
            try:
                return convert_to(self.data.get(key, None))
            except ValueError:
                _LOGGER.warning("Value %s can't be converted to %s", key, convert_to)
                return None

        _LOGGER.warning("Value %s is missing in API response", key)
        return None
