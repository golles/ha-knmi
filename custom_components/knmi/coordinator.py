"""DataUpdateCoordinator for knmi."""
import logging
from typing import Any, Callable

from homeassistant.config_entries import ConfigEntry
from homeassistant.core import HomeAssistant
from homeassistant.helpers.entity import DeviceInfo
from homeassistant.helpers.update_coordinator import DataUpdateCoordinator, UpdateFailed

from .api import KnmiApiClient
from .const import DOMAIN, SCAN_INTERVAL

_LOGGER: logging.Logger = logging.getLogger(__package__)


class KnmiDataUpdateCoordinator(DataUpdateCoordinator):
    """Class to manage fetching data from the API."""

    config_entry: ConfigEntry

    def __init__(
        self, hass: HomeAssistant, client: KnmiApiClient, device_info: DeviceInfo
    ) -> None:
        """Initialize."""
        self.api = client
        self.device_info = device_info

        super().__init__(
            hass=hass, logger=_LOGGER, name=DOMAIN, update_interval=SCAN_INTERVAL
        )

    async def _async_update_data(self) -> dict[str, Any]:
        """Update data via library."""
        try:
            return await self.api.async_get_data()
        except Exception as exception:
            _LOGGER.error("Update failed! - %s", exception)
            raise UpdateFailed() from exception

    def get_value(
        self, key: str, convert_to: Callable = str
    ) -> float | int | str | None:
        """Get a value from the retrieved data and convert to given type"""
        if self.data and key in self.data:
            if self.data.get(key, None) == "" and key != "alarmtxt":
                _LOGGER.warning("Value %s is empty in API response", key)
                return ""  # Leave empty, eg. warning attribute can be an empty string.

            try:
                return convert_to(self.data.get(key, None))
            except ValueError:
                _LOGGER.warning("Value %s can't be converted to %s", key, convert_to)
                return None

        _LOGGER.warning("Value %s is missing in API response", key)
        return None
