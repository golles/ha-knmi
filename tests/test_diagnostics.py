"""Test diagnostics."""

import pytest
from homeassistant.components.diagnostics import REDACTED
from homeassistant.core import HomeAssistant
from pytest_homeassistant_custom_component.components.diagnostics import get_diagnostics_for_config_entry
from pytest_homeassistant_custom_component.typing import ClientSessionGenerator

from custom_components.knmi.const import DOMAIN
from custom_components.knmi.diagnostics import TO_REDACT

from . import setup_integration, unload_integration


@pytest.mark.usefixtures("mocked_data")
async def test_config_entry_diagnostics(hass: HomeAssistant, hass_client: ClientSessionGenerator) -> None:
    """Test config entry diagnostics."""
    config_entry = await setup_integration(hass)

    result = await get_diagnostics_for_config_entry(hass, hass_client, config_entry)

    assert result["config_entry"]["entry_id"] == "test_entry"
    assert result["config_entry"]["domain"] == DOMAIN

    for key in TO_REDACT:
        assert result["config_entry"]["data"][key] == REDACTED

    assert result["data"]["live"]["city"] == "Purmerend"

    await unload_integration(hass, config_entry)
