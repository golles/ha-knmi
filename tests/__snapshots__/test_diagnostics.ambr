# serializer version: 1
# name: test_config_entry_diagnostics
  dict({
    'config_entry': dict({
      'data': dict({
        'api_key': '**REDACTED**',
        'latitude': '**REDACTED**',
        'longitude': '**REDACTED**',
        'name': 'Home',
      }),
      'disabled_by': None,
      'domain': 'knmi',
      'entry_id': 'test',
      'minor_version': 1,
      'options': dict({
      }),
      'pref_disable_new_entities': False,
      'pref_disable_polling': False,
      'source': 'user',
      'title': 'Mock Title',
      'unique_id': None,
      'version': 2,
    }),
    'data': dict({
      'api': list([
        dict({
          'bron': 'Bron: Weerdata KNMI/NOAA via Weerlive.nl',
          'max_verz': 300,
          'rest_verz': 132,
        }),
      ]),
      'liveweer': list([
        dict({
          'alarm': 0,
          'dauwp': 10.1,
          'gr': 0,
          'gtemp': 6.8,
          'image': 'wolkennacht',
          'ldmmhg': 761,
          'lkop': 'Vanavond (zeer) zware windstoten',
          'ltekst': 'De eerstkomende uren zijn er geen waarschuwingen van kracht. Vanavond komen er (zeer) zware windstoten voor. Landinwaarts tot 90 km/u, aan de kust tot 110 km/u. De wind komt uit een zuidwestelijke richting. Verkeer en buitenactiviteiten kunnen hinder ondervinden.',
          'luchtd': 1015.03,
          'lv': 97,
          'plaats': 'Purmerend',
          'samenv': 'Licht bewolkt',
          'sunder': '17:51',
          'sup': '07:57',
          'temp': 10.5,
          'time': '14-02-2024 22:08:03',
          'timestamp': 1707944883,
          'verw': 'Bewolkt en perioden met regen. Morgen in de middag droog en zeer zacht',
          'windbft': 5,
          'windkmh': 29.1,
          'windknp': 15.7,
          'windms': 8.08,
          'windr': 'WZW',
          'windrgr': 226,
          'wrsch_g': '22-02-2024 18:00',
          'wrsch_gc': 'geel',
          'wrsch_gts': 1708621200,
          'wrschklr': 'groen',
          'zicht': 6990,
        }),
      ]),
      'uur_verw': list([
        dict({
          'gr': 0,
          'image': 'nachtbewolkt',
          'neersl': 0,
          'temp': 10,
          'timestamp': 1707948000,
          'uur': '14-02-2024 23:00',
          'windbft': 4,
          'windkmh': 21,
          'windknp': 12,
          'windms': 6,
          'windr': 'ZW',
          'windrgr': 231,
        }),
        dict({
          'gr': 0,
          'image': 'nachtbewolkt',
          'neersl': 0,
          'temp': 10,
          'timestamp': 1707951600,
          'uur': '15-02-2024 00:00',
          'windbft': 3,
          'windkmh': 18,
          'windknp': 10,
          'windms': 5,
          'windr': 'ZW',
          'windrgr': 233,
        }),
        dict({
          'gr': 0,
          'image': 'regen',
          'neersl': 0.1,
          'temp': 10,
          'timestamp': 1707955200,
          'uur': '15-02-2024 01:00',
          'windbft': 3,
          'windkmh': 18,
          'windknp': 10,
          'windms': 5,
          'windr': 'ZW',
          'windrgr': 232,
        }),
        dict({
          'gr': 0,
          'image': 'bewolkt',
          'neersl': 0,
          'temp': 10,
          'timestamp': 1707958800,
          'uur': '15-02-2024 02:00',
          'windbft': 3,
          'windkmh': 18,
          'windknp': 10,
          'windms': 5,
          'windr': 'ZW',
          'windrgr': 226,
        }),
        dict({
          'gr': 0,
          'image': 'bewolkt',
          'neersl': 0,
          'temp': 10,
          'timestamp': 1707962400,
          'uur': '15-02-2024 03:00',
          'windbft': 3,
          'windkmh': 18,
          'windknp': 10,
          'windms': 5,
          'windr': 'ZW',
          'windrgr': 224,
        }),
        dict({
          'gr': 0,
          'image': 'bewolkt',
          'neersl': 0,
          'temp': 10,
          'timestamp': 1707966000,
          'uur': '15-02-2024 04:00',
          'windbft': 3,
          'windkmh': 18,
          'windknp': 10,
          'windms': 5,
          'windr': 'ZW',
          'windrgr': 219,
        }),
        dict({
          'gr': 0,
          'image': 'bewolkt',
          'neersl': 0,
          'temp': 10,
          'timestamp': 1707969600,
          'uur': '15-02-2024 05:00',
          'windbft': 3,
          'windkmh': 14,
          'windknp': 8,
          'windms': 4,
          'windr': 'Z',
          'windrgr': 203,
        }),
        dict({
          'gr': 0,
          'image': 'regen',
          'neersl': 0.2,
          'temp': 10,
          'timestamp': 1707973200,
          'uur': '15-02-2024 06:00',
          'windbft': 3,
          'windkmh': 14,
          'windknp': 8,
          'windms': 4,
          'windr': 'Z',
          'windrgr': 190,
        }),
        dict({
          'gr': 0,
          'image': 'regen',
          'neersl': 0.9,
          'temp': 10,
          'timestamp': 1707976800,
          'uur': '15-02-2024 07:00',
          'windbft': 3,
          'windkmh': 14,
          'windknp': 8,
          'windms': 4,
          'windr': 'Z',
          'windrgr': 196,
        }),
        dict({
          'gr': 0,
          'image': 'regen',
          'neersl': 1,
          'temp': 11,
          'timestamp': 1707980400,
          'uur': '15-02-2024 08:00',
          'windbft': 3,
          'windkmh': 18,
          'windknp': 10,
          'windms': 5,
          'windr': 'Z',
          'windrgr': 210,
        }),
        dict({
          'gr': 3,
          'image': 'regen',
          'neersl': 0.1,
          'temp': 11,
          'timestamp': 1707984000,
          'uur': '15-02-2024 09:00',
          'windbft': 3,
          'windkmh': 14,
          'windknp': 8,
          'windms': 4,
          'windr': 'ZW',
          'windrgr': 218,
        }),
        dict({
          'gr': 6,
          'image': 'bewolkt',
          'neersl': 0,
          'temp': 11,
          'timestamp': 1707987600,
          'uur': '15-02-2024 10:00',
          'windbft': 3,
          'windkmh': 14,
          'windknp': 8,
          'windms': 4,
          'windr': 'ZW',
          'windrgr': 217,
        }),
        dict({
          'gr': 11,
          'image': 'regen',
          'neersl': 0.1,
          'temp': 11,
          'timestamp': 1707991200,
          'uur': '15-02-2024 11:00',
          'windbft': 3,
          'windkmh': 14,
          'windknp': 8,
          'windms': 4,
          'windr': 'ZW',
          'windrgr': 214,
        }),
        dict({
          'gr': 30,
          'image': 'bewolkt',
          'neersl': 0,
          'temp': 12,
          'timestamp': 1707994800,
          'uur': '15-02-2024 12:00',
          'windbft': 3,
          'windkmh': 14,
          'windknp': 8,
          'windms': 4,
          'windr': 'Z',
          'windrgr': 206,
        }),
        dict({
          'gr': 172,
          'image': 'bewolkt',
          'neersl': 0,
          'temp': 13,
          'timestamp': 1707998400,
          'uur': '15-02-2024 13:00',
          'windbft': 3,
          'windkmh': 14,
          'windknp': 8,
          'windms': 4,
          'windr': 'Z',
          'windrgr': 211,
        }),
        dict({
          'gr': 161,
          'image': 'bewolkt',
          'neersl': 0,
          'temp': 14,
          'timestamp': 1708002000,
          'uur': '15-02-2024 14:00',
          'windbft': 3,
          'windkmh': 14,
          'windknp': 8,
          'windms': 4,
          'windr': 'Z',
          'windrgr': 200,
        }),
        dict({
          'gr': 58,
          'image': 'bewolkt',
          'neersl': 0,
          'temp': 14,
          'timestamp': 1708005600,
          'uur': '15-02-2024 15:00',
          'windbft': 2,
          'windkmh': 10,
          'windknp': 6,
          'windms': 3,
          'windr': 'Z',
          'windrgr': 179,
        }),
        dict({
          'gr': 114,
          'image': 'bewolkt',
          'neersl': 0,
          'temp': 14,
          'timestamp': 1708009200,
          'uur': '15-02-2024 16:00',
          'windbft': 2,
          'windkmh': 10,
          'windknp': 6,
          'windms': 3,
          'windr': 'ZO',
          'windrgr': 167,
        }),
        dict({
          'gr': 72,
          'image': 'bewolkt',
          'neersl': 0,
          'temp': 14,
          'timestamp': 1708012800,
          'uur': '15-02-2024 17:00',
          'windbft': 3,
          'windkmh': 14,
          'windknp': 8,
          'windms': 4,
          'windr': 'Z',
          'windrgr': 168,
        }),
        dict({
          'gr': 14,
          'image': 'bewolkt',
          'neersl': 0,
          'temp': 13,
          'timestamp': 1708016400,
          'uur': '15-02-2024 18:00',
          'windbft': 3,
          'windkmh': 14,
          'windknp': 8,
          'windms': 4,
          'windr': 'ZO',
          'windrgr': 156,
        }),
        dict({
          'gr': 0,
          'image': 'bewolkt',
          'neersl': 0,
          'temp': 12,
          'timestamp': 1708020000,
          'uur': '15-02-2024 19:00',
          'windbft': 3,
          'windkmh': 14,
          'windknp': 8,
          'windms': 4,
          'windr': 'ZO',
          'windrgr': 155,
        }),
        dict({
          'gr': 0,
          'image': 'bewolkt',
          'neersl': 0,
          'temp': 12,
          'timestamp': 1708023600,
          'uur': '15-02-2024 20:00',
          'windbft': 3,
          'windkmh': 18,
          'windknp': 10,
          'windms': 5,
          'windr': 'ZO',
          'windrgr': 152,
        }),
        dict({
          'gr': 0,
          'image': 'bewolkt',
          'neersl': 0,
          'temp': 12,
          'timestamp': 1708027200,
          'uur': '15-02-2024 21:00',
          'windbft': 3,
          'windkmh': 18,
          'windknp': 10,
          'windms': 5,
          'windr': 'ZO',
          'windrgr': 155,
        }),
        dict({
          'gr': 0,
          'image': 'bewolkt',
          'neersl': 0,
          'temp': 12,
          'timestamp': 1708030800,
          'uur': '15-02-2024 22:00',
          'windbft': 3,
          'windkmh': 18,
          'windknp': 10,
          'windms': 5,
          'windr': 'ZO',
          'windrgr': 162,
        }),
      ]),
      'wk_verw': list([
        dict({
          'dag': '14-02-2024',
          'image': 'bewolkt',
          'max_temp': 10,
          'min_temp': 10,
          'neersl_perc_dag': 0,
          'windbft': 4,
          'windkmh': 25,
          'windknp': 14,
          'windms': 7,
          'windr': 'ZW',
          'windrgr': 221,
          'zond_perc_dag': 0,
        }),
        dict({
          'dag': '15-02-2024',
          'image': 'bewolkt',
          'max_temp': 12,
          'min_temp': 10,
          'neersl_perc_dag': 10,
          'windbft': 3,
          'windkmh': 18,
          'windknp': 10,
          'windms': 5,
          'windr': 'Z',
          'windrgr': 184,
          'zond_perc_dag': 8,
        }),
        dict({
          'dag': '16-02-2024',
          'image': 'buien',
          'max_temp': 10,
          'min_temp': 9,
          'neersl_perc_dag': 40,
          'windbft': 3,
          'windkmh': 18,
          'windknp': 10,
          'windms': 5,
          'windr': 'Z',
          'windrgr': 199,
          'zond_perc_dag': 14,
        }),
        dict({
          'dag': '17-02-2024',
          'image': 'halfbewolkt',
          'max_temp': 8,
          'min_temp': 6,
          'neersl_perc_dag': 20,
          'windbft': 3,
          'windkmh': 18,
          'windknp': 10,
          'windms': 5,
          'windr': 'ZW',
          'windrgr': 228,
          'zond_perc_dag': 46,
        }),
        dict({
          'dag': '18-02-2024',
          'image': 'bewolkt',
          'max_temp': 8,
          'min_temp': 7,
          'neersl_perc_dag': 10,
          'windbft': 3,
          'windkmh': 18,
          'windknp': 10,
          'windms': 5,
          'windr': 'Z',
          'windrgr': 210,
          'zond_perc_dag': 0,
        }),
      ]),
    }),
    'response_text': '''
      {
        "liveweer": [
          {
            "plaats": "Purmerend",
            "timestamp": 1707944883,
            "time": "14-02-2024 22:08:03",
            "temp": 10.5,
            "gtemp": 6.8,
            "samenv": "Licht bewolkt",
            "lv": 97,
            "windr": "WZW",
            "windrgr": 226,
            "windms": 8.08,
            "windbft": 5,
            "windknp": 15.7,
            "windkmh": 29.1,
            "luchtd": 1015.03,
            "ldmmhg": 761,
            "dauwp": 10.1,
            "zicht": 6990,
            "gr": 0,
            "verw": "Bewolkt en perioden met regen. Morgen in de middag droog en zeer zacht",
            "sup": "07:57",
            "sunder": "17:51",
            "image": "wolkennacht",
            "alarm": 0,
            "lkop": "Vanavond (zeer) zware windstoten",
            "ltekst": "De eerstkomende uren zijn er geen waarschuwingen van kracht. Vanavond komen er (zeer) zware windstoten voor. Landinwaarts tot 90 km/u, aan de kust tot 110 km/u. De wind komt uit een zuidwestelijke richting. Verkeer en buitenactiviteiten kunnen hinder ondervinden.",
            "wrschklr": "groen",
            "wrsch_g": "22-02-2024 18:00",
            "wrsch_gts": 1708621200,
            "wrsch_gc": "geel"
          }
        ],
        "wk_verw": [
          {
            "dag": "14-02-2024",
            "image": "bewolkt",
            "max_temp": 10,
            "min_temp": 10,
            "windbft": 4,
            "windkmh": 25,
            "windknp": 14,
            "windms": 7,
            "windrgr": 221,
            "windr": "ZW",
            "neersl_perc_dag": 0,
            "zond_perc_dag": 0
          },
          {
            "dag": "15-02-2024",
            "image": "bewolkt",
            "max_temp": 12,
            "min_temp": 10,
            "windbft": 3,
            "windkmh": 18,
            "windknp": 10,
            "windms": 5,
            "windrgr": 184,
            "windr": "Z",
            "neersl_perc_dag": 10,
            "zond_perc_dag": 8
          },
          {
            "dag": "16-02-2024",
            "image": "buien",
            "max_temp": 10,
            "min_temp": 9,
            "windbft": 3,
            "windkmh": 18,
            "windknp": 10,
            "windms": 5,
            "windrgr": 199,
            "windr": "Z",
            "neersl_perc_dag": 40,
            "zond_perc_dag": 14
          },
          {
            "dag": "17-02-2024",
            "image": "halfbewolkt",
            "max_temp": 8,
            "min_temp": 6,
            "windbft": 3,
            "windkmh": 18,
            "windknp": 10,
            "windms": 5,
            "windrgr": 228,
            "windr": "ZW",
            "neersl_perc_dag": 20,
            "zond_perc_dag": 46
          },
          {
            "dag": "18-02-2024",
            "image": "bewolkt",
            "max_temp": 8,
            "min_temp": 7,
            "windbft": 3,
            "windkmh": 18,
            "windknp": 10,
            "windms": 5,
            "windrgr": 210,
            "windr": "Z",
            "neersl_perc_dag": 10,
            "zond_perc_dag": 0
          }
        ],
        "uur_verw": [
          {
            "uur": "14-02-2024 23:00",
            "timestamp": 1707948000,
            "image": "nachtbewolkt",
            "temp": 10,
            "windbft": 4,
            "windkmh": 21,
            "windknp": 12,
            "windms": 6,
            "windrgr": 231,
            "windr": "ZW",
            "neersl": 0,
            "gr": 0
          },
          {
            "uur": "15-02-2024 00:00",
            "timestamp": 1707951600,
            "image": "nachtbewolkt",
            "temp": 10,
            "windbft": 3,
            "windkmh": 18,
            "windknp": 10,
            "windms": 5,
            "windrgr": 233,
            "windr": "ZW",
            "neersl": 0,
            "gr": 0
          },
          {
            "uur": "15-02-2024 01:00",
            "timestamp": 1707955200,
            "image": "regen",
            "temp": 10,
            "windbft": 3,
            "windkmh": 18,
            "windknp": 10,
            "windms": 5,
            "windrgr": 232,
            "windr": "ZW",
            "neersl": 0.1,
            "gr": 0
          },
          {
            "uur": "15-02-2024 02:00",
            "timestamp": 1707958800,
            "image": "bewolkt",
            "temp": 10,
            "windbft": 3,
            "windkmh": 18,
            "windknp": 10,
            "windms": 5,
            "windrgr": 226,
            "windr": "ZW",
            "neersl": 0,
            "gr": 0
          },
          {
            "uur": "15-02-2024 03:00",
            "timestamp": 1707962400,
            "image": "bewolkt",
            "temp": 10,
            "windbft": 3,
            "windkmh": 18,
            "windknp": 10,
            "windms": 5,
            "windrgr": 224,
            "windr": "ZW",
            "neersl": 0,
            "gr": 0
          },
          {
            "uur": "15-02-2024 04:00",
            "timestamp": 1707966000,
            "image": "bewolkt",
            "temp": 10,
            "windbft": 3,
            "windkmh": 18,
            "windknp": 10,
            "windms": 5,
            "windrgr": 219,
            "windr": "ZW",
            "neersl": 0,
            "gr": 0
          },
          {
            "uur": "15-02-2024 05:00",
            "timestamp": 1707969600,
            "image": "bewolkt",
            "temp": 10,
            "windbft": 3,
            "windkmh": 14,
            "windknp": 8,
            "windms": 4,
            "windrgr": 203,
            "windr": "Z",
            "neersl": 0,
            "gr": 0
          },
          {
            "uur": "15-02-2024 06:00",
            "timestamp": 1707973200,
            "image": "regen",
            "temp": 10,
            "windbft": 3,
            "windkmh": 14,
            "windknp": 8,
            "windms": 4,
            "windrgr": 190,
            "windr": "Z",
            "neersl": 0.2,
            "gr": 0
          },
          {
            "uur": "15-02-2024 07:00",
            "timestamp": 1707976800,
            "image": "regen",
            "temp": 10,
            "windbft": 3,
            "windkmh": 14,
            "windknp": 8,
            "windms": 4,
            "windrgr": 196,
            "windr": "Z",
            "neersl": 0.9,
            "gr": 0
          },
          {
            "uur": "15-02-2024 08:00",
            "timestamp": 1707980400,
            "image": "regen",
            "temp": 11,
            "windbft": 3,
            "windkmh": 18,
            "windknp": 10,
            "windms": 5,
            "windrgr": 210,
            "windr": "Z",
            "neersl": 1,
            "gr": 0
          },
          {
            "uur": "15-02-2024 09:00",
            "timestamp": 1707984000,
            "image": "regen",
            "temp": 11,
            "windbft": 3,
            "windkmh": 14,
            "windknp": 8,
            "windms": 4,
            "windrgr": 218,
            "windr": "ZW",
            "neersl": 0.1,
            "gr": 3
          },
          {
            "uur": "15-02-2024 10:00",
            "timestamp": 1707987600,
            "image": "bewolkt",
            "temp": 11,
            "windbft": 3,
            "windkmh": 14,
            "windknp": 8,
            "windms": 4,
            "windrgr": 217,
            "windr": "ZW",
            "neersl": 0,
            "gr": 6
          },
          {
            "uur": "15-02-2024 11:00",
            "timestamp": 1707991200,
            "image": "regen",
            "temp": 11,
            "windbft": 3,
            "windkmh": 14,
            "windknp": 8,
            "windms": 4,
            "windrgr": 214,
            "windr": "ZW",
            "neersl": 0.1,
            "gr": 11
          },
          {
            "uur": "15-02-2024 12:00",
            "timestamp": 1707994800,
            "image": "bewolkt",
            "temp": 12,
            "windbft": 3,
            "windkmh": 14,
            "windknp": 8,
            "windms": 4,
            "windrgr": 206,
            "windr": "Z",
            "neersl": 0,
            "gr": 30
          },
          {
            "uur": "15-02-2024 13:00",
            "timestamp": 1707998400,
            "image": "bewolkt",
            "temp": 13,
            "windbft": 3,
            "windkmh": 14,
            "windknp": 8,
            "windms": 4,
            "windrgr": 211,
            "windr": "Z",
            "neersl": 0,
            "gr": 172
          },
          {
            "uur": "15-02-2024 14:00",
            "timestamp": 1708002000,
            "image": "bewolkt",
            "temp": 14,
            "windbft": 3,
            "windkmh": 14,
            "windknp": 8,
            "windms": 4,
            "windrgr": 200,
            "windr": "Z",
            "neersl": 0,
            "gr": 161
          },
          {
            "uur": "15-02-2024 15:00",
            "timestamp": 1708005600,
            "image": "bewolkt",
            "temp": 14,
            "windbft": 2,
            "windkmh": 10,
            "windknp": 6,
            "windms": 3,
            "windrgr": 179,
            "windr": "Z",
            "neersl": 0,
            "gr": 58
          },
          {
            "uur": "15-02-2024 16:00",
            "timestamp": 1708009200,
            "image": "bewolkt",
            "temp": 14,
            "windbft": 2,
            "windkmh": 10,
            "windknp": 6,
            "windms": 3,
            "windrgr": 167,
            "windr": "ZO",
            "neersl": 0,
            "gr": 114
          },
          {
            "uur": "15-02-2024 17:00",
            "timestamp": 1708012800,
            "image": "bewolkt",
            "temp": 14,
            "windbft": 3,
            "windkmh": 14,
            "windknp": 8,
            "windms": 4,
            "windrgr": 168,
            "windr": "Z",
            "neersl": 0,
            "gr": 72
          },
          {
            "uur": "15-02-2024 18:00",
            "timestamp": 1708016400,
            "image": "bewolkt",
            "temp": 13,
            "windbft": 3,
            "windkmh": 14,
            "windknp": 8,
            "windms": 4,
            "windrgr": 156,
            "windr": "ZO",
            "neersl": 0,
            "gr": 14
          },
          {
            "uur": "15-02-2024 19:00",
            "timestamp": 1708020000,
            "image": "bewolkt",
            "temp": 12,
            "windbft": 3,
            "windkmh": 14,
            "windknp": 8,
            "windms": 4,
            "windrgr": 155,
            "windr": "ZO",
            "neersl": 0,
            "gr": 0
          },
          {
            "uur": "15-02-2024 20:00",
            "timestamp": 1708023600,
            "image": "bewolkt",
            "temp": 12,
            "windbft": 3,
            "windkmh": 18,
            "windknp": 10,
            "windms": 5,
            "windrgr": 152,
            "windr": "ZO",
            "neersl": 0,
            "gr": 0
          },
          {
            "uur": "15-02-2024 21:00",
            "timestamp": 1708027200,
            "image": "bewolkt",
            "temp": 12,
            "windbft": 3,
            "windkmh": 18,
            "windknp": 10,
            "windms": 5,
            "windrgr": 155,
            "windr": "ZO",
            "neersl": 0,
            "gr": 0
          },
          {
            "uur": "15-02-2024 22:00",
            "timestamp": 1708030800,
            "image": "bewolkt",
            "temp": 12,
            "windbft": 3,
            "windkmh": 18,
            "windknp": 10,
            "windms": 5,
            "windrgr": 162,
            "windr": "ZO",
            "neersl": 0,
            "gr": 0
          }
        ],
        "api": [
          {
            "bron": "Bron: Weerdata KNMI/NOAA via Weerlive.nl",
            "max_verz": 300,
            "rest_verz": 132
          }
        ]
      }
  
    ''',
  })
# ---
